packages/driver/src/fetchConn.ts

```diff
@@ -136,6 +145,21 @@ class BaseFetchConnection extends BaseRawConnection {
     this.__sendData(data);
   }
 
+  async fetch(...args: Parameters<BaseRawConnection["fetch"]>) {
+    const protoVer = this.protocolVersion;
+    try {
+      return await super.fetch(...args);
+    } catch (err: unknown) {
+      if (
+        err instanceof BinaryProtocolError &&
+        !versionEqual(protoVer, this.protocolVersion)
+      ) {
+        return await super.fetch(...args);
+      }
```

Comment:
In the normal tcp connection the protocol version is negotiated during the connection handshake, so the driver knows the supported proto ver to use when it sends the first parse/execute message (which are now different for v3). But for fetch conn, we only know the servers supported proto ver after sending the first message, so the first message is sent using proto v3 format, and if the server only supports v2 it will return a BinaryProtocolError as it tries to decode the message as though it's in v2 format. So this code is to workaround that by retrying with the correct proto ver, if the proto ver used the send the first message doesn't match the version the server returned on the response and that caused a BinaryProtocolError.

Reply:
I see, so `this.protocolVersion` is mutated after doing the first call, so if the original call was made with v3 but the server uses v2, `protoVer` will be v3 but `this.protocolVersion` will be v2?

Reply:
We didn't have the same issue for v1 -> v2, as that only changed the typedesc format, so the client only needs the correct protocol version to decode the response, which it gets along with that response.

Reply:
I should probably add a comment in the code 😄.

Reply:
> I see, so `this.protocolVersion` is mutated after doing the first call, so if the original call was made with v3 but the server uses v2, `protoVer` will be v3 but `this.protocolVersion` will be v2?

Yes: https://github.com/edgedb/edgedb-js/blob/master/packages/driver/src/fetchConn.ts#L109-L113

Reply:
Note to self/maintainers: Not a blocker, but I wonder if `ensureConnected` can do this protocol negotiation, too, to avoid the overhead of two roundtrips on the first connection? It wouldn't really _save_ a roundtrip, but it might make query times more predictable. We can introduce this in a future update as an optimization while keeping the existing retry-style logic.